worker/instancepoller/worker.go

```diff
@@ -85,57 +180,319 @@ func (u *updaterWorker) Wait() error {
+			continue // we shouldn't poll this entry yet
+		}
+
+		if err := u.resolveInstanceID(entry); err != nil {
+			if params.IsCodeNotProvisioned(err) {
+				// machine not provisioned yet; bump its poll
+				// interval and re-try later (or as soon as we
+				// get a change for the machine)
+				entry.bumpShortPollInterval(u.config.Clock)
+				continue
+			}
+			return errors.Trace(err)
+		}
+
+		instList = append(instList, entry.instanceID)
+	}
+
+	if len(instList) == 0 {
+		return nil
+	}
+
+	infoList, err := u.config.Environ.Instances(u.callContext, instList)
+	if err != nil && err != environs.ErrPartialInstances {
+		return errors.Trace(err)
+	}
+	for idx, info := range infoList {
+		// No details found for this instance. This most probably means
+		// that the unit has been killed and we haven't been notified
+		// yet. Log the error and keep going.
+		if info == nil {
+			u.config.Logger.Warningf("unable to retrieve instance information for instance: %q", instList[idx])
+			continue
+		}
+
+		entry := u.instanceIDToGroupEntry[instList[idx]]
+		providerStatus, err := u.processProviderInfo(entry, info)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		machineStatus, err := entry.m.Status()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		u.maybeSwitchPollGroup(groupType, entry, providerStatus, status.Status(machineStatus.Status))
+	}
+
+	return nil
 }
 
-// instanceInfo is part of the machineContext interface.
-func (u *updaterWorker) instanceInfo(id instance.Id) (instanceInfo, error) {
-	return u.aggregator.instanceInfo(id)
+func (u *updaterWorker) resolveInstanceID(entry *pollGroupEntry) error {
+	if entry.instanceID != "" {
+		return nil // already resolved
+	}
+
+	instID, err := entry.m.InstanceId()
+	if err != nil {
+		return errors.Annotate(err, "cannot get machine's instance ID")
+	}
+
+	entry.instanceID = instID
+	u.instanceIDToGroupEntry[instID] = entry
+	return nil
 }
 
-// kill is part of the lifetimeContext interface.
-func (u *updaterWorker) kill(err error) {
-	u.catacomb.Kill(err)
+func (u *updaterWorker) processProviderInfo(entry *pollGroupEntry, info instances.Instance) (status.Status, error) {
+	curStatus, err := entry.m.InstanceStatus()
+	if err != nil {
+		// This should never occur since the machine is provisioned. If
+		// it does occur, report an unknown status to move the machine to
+		// the short poll group.
+		u.config.Logger.Warningf("cannot get current instance status for machine %v (instance ID %q): %v", entry.m.Id(), entry.instanceID, err)
+		return status.Unknown, nil
+	}
+
+	// Check for status changes
+	providerStatus := info.Status(u.callContext)
+	curInstStatus := instance.Status{
+		Status:  status.Status(curStatus.Status),
+		Message: curStatus.Info,
+	}
+
+	if providerStatus != curInstStatus {
+		u.config.Logger.Infof("machine %q (instance ID %q) instance status changed from %q to %q", entry.m.Id(), entry.instanceID, curInstStatus, providerStatus)
```

Comment:
I wonder if this should be Info or Debug. Is it something we want to display by default in logs?